### PR TITLE
feat(cli): add doctor diagnostics and help snapshots

### LINKED ISSUE #13: [cli] Command coverage and dsar doctor diagnostic command
## Why
CLIs feel mature or abandoned largely based on `--help` quality and whether commands crash on edge cases. It is usually the first thing an evaluator touches. Today 11 commands exist with uneven coverage.

## Scope
- Integration test per command in `packages/cli/src/commands/**` (audit, notifications, policies, registry, requests, retention, subjects, system, webhooks + sub-commands)
- Golden snapshots of `--help` output for every command
- New `dsar doctor` command that validates:
  - Runtime config is loadable
  - DB connection works + migrations are current
  - Each configured adapter can initialize (storage, inbound, outbound)
  - Auth provider (Unkey or local) is reachable
- Exit codes documented and tested

**Priority:** P1 - beta blocker

### PR DESCRIPTION
**Pull request by devcool20**
## Linked Issue

Closes #13.

Implemented:
- `dsar doctor` config/runtime/auth/persistence reachability checks
- command-specific and command-group help
- help snapshots and regression coverage
- sanitized doctor diagnostics output
- readable default doctor text output with `--output json` retained for automation

Completed in follow-up commits on this PR (maintainer update):
- operator/service-scoped `GET /status/diagnostics` backend endpoint exposing schema migration freshness and per-adapter health
- `dsar doctor` now fails on stale migrations or down adapters, warns on degraded adapters, and keeps graceful warn/skip behavior against older backends
- changeset added (`dsar`: minor)

## Description

Adds an initial `dsar doctor` command for CLI/runtime diagnostics and expands CLI help coverage.

This includes:
- API URL validation
- runtime `/status` reachability check
- authenticated protected-request probe when a token is configured
- warning/skipped diagnostics for migration freshness and adapter health when the backend does not expose those surfaces yet
- command-specific help output via `dsar <command> --help`
- grouped help output via `dsar requests --help`
- golden help snapshots for every registered CLI command
- exit-code documentation

With the diagnostics endpoint added in the follow-up commits, this PR now fully closes #13.

##Screenshots for validation

Screenshot 2026-05-18 222129
Screenshot 2026-05-18 222057
Screenshot 2026-05-18 222013

## Validation

```powershell
bun --cwd packages/cli vitest run
bun run --cwd packages/cli typecheck
bun x ultracite check packages/cli/src/config.ts packages/cli/src/runtime.ts packages/cli/src/types.ts packages/cli/src/commands/doctor.ts packages/cli/src/commands/system.ts packages/cli/test/doctor.test.ts packages/cli/test/help.test.ts packages/cli/test/e2e/parity-guard.e2e.test.ts packages/cli/README.md docs/reference/developer/cli.mdx
bun run check:docs-lint
bun run check:tsdoc
git diff --check

### CODE REVIEW

**Comment by devcool20** [comment]
Updated after review:

- Sanitized the authenticated `/requests` doctor probe so protected response payloads are no longer emitted in diagnostics.
- Fixed targeted `-h`/`--help` handling by stripping help flags before help target resolution.
- Updated help tests to capture full stdout.
- Added command-owned doctor text formatting while preserving `--output json`.
- Replaced the runtime’s hardcoded doctor special case with command metadata.
- Moved doctor coverage under `test/e2e`.

Validation passed locally:
- `bun --cwd packages/cli vitest run`
- `bun run --cwd packages/cli typecheck`
- `bun x ultracite check ...`
- `bun run check:docs-lint`
- `bun run check:tsdoc`
- `git diff --check`

**Comment by KayleeWilliams** [comment]
Maintainer review + update: the original scope was solid (doctor command, help snapshots, exit codes, honest warn/skip for missing backend surfaces), and the diff is clean from a supply-chain standpoint (no new deps, no workflow changes). Since we only merge full fixes, I pushed follow-up commits completing issue #13:

- **`GET /status/diagnostics`** (operator/service principals only): schema migration freshness via a new read-only `readMigrationStatus`/`migrationStatus` surface in `@dsar/persistence`, plus per-adapter `healthCheck()`/`diagnostics()` enumeration from the adapter registry (sanitized output).
- **doctor wiring**: `persistence.migrations` now passes/fails on actual freshness (listing missing migrations on failure); `adapters.health` passes when all healthy, warns on degraded, fails on down; both degrade gracefully (warn/skip) against older backends without the endpoint.
- OpenAPI group + snapshot, backend route tests (authz + shape + stale-migration case), doctor e2e updates, docs/README updates, and a `dsar` minor changeset.

Verified: full backend (100), cli (159), and persistence (23) suites pass; ultracite, tsdoc, error-docs, docs-lint, alpha-warning all clean. Note for merge sequencing: this touches the same CLI parity/help files as #44, so whichever lands second needs a trivial rebase.

File: packages/cli/src/runtime.ts [hunk 1]
```diff
@@ -20,58 +20,78 @@ const asEnvelope = (
...
-		);
+				: undefined,
+			ok: false,
+		});
 	}
 	if (input.ok) {
 		return JSON.stringify(input.data, null, 2);
 	}
+	if ("data" in input) {
+		return JSON.stringify(input.data, null, 2);
```
[the remark this answers is not in the document]
Comment (KayleeWilliams) [reply packages/cli/src/runtime.ts]: Addressed — `CommandDefinition.formatTextResult` was added and `dsar doctor` now renders a per-check text report (`[pass] runtime.status — ...` plus a summary line) in the default text mode for both passing and failing runs; JSON remains available via `--output json`. Covered by the "renders readable text output by default" e2e case.

File: packages/cli/src/runtime.ts [hunk 2]
```diff
@@ -81,9 +101,78 @@ const renderHelp = (): string => {
...
+	}
+	return renderCommandGroupHelp(commandTokens);
+};
+
 const isHelpRequest = (argv: readonly string[]): boolean =>
 	argv.includes("--help") || argv.includes("-h");
 
+const isDoctorRequest = (argv: readonly string[]): boolean => {
+	const { commandTokens } = parseCliArguments(argv);
+	return commandTokens.length === 1 && commandTokens[0] === "doctor";
```
[the remark this answers is not in the document]
Comment (KayleeWilliams) [reply packages/cli/src/runtime.ts]: Addressed — the literal `commandTokens[0] === \"doctor\"` branch was replaced by a declarative `CommandDefinition.allowMissingApiUrl` flag: the runtime resolves the command first and passes `matched?.command.allowMissingApiUrl === true` into `parseCliInput`, so any future command can opt in without touching the runtime.

File: packages/cli/test/doctor.test.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
+import { describe, expect, it } from "@effect/vitest";
+
+import { runCli } from "#src/runtime";
+
+import { E2E_API_TOKEN, E2E_API_URL, makeRuntimeFetch } from "./e2e/harness";
```
[the remark this answers is not in the document]
Comment (KayleeWilliams) [reply packages/cli/test/doctor.test.ts]: Addressed — the test now lives at `packages/cli/test/e2e/doctor.e2e.test.ts`, matching the e2e naming convention of its harness-backed siblings.

File: packages/cli/src/commands/doctor.ts [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+		return value;
+	}
+	if (isRecord(value) && isDiagnosticsData(value.data)) {
+		return value.data;
+	}
+	return undefined;
+};
+
+const isUnavailableDiagnosticsError = (error: unknown): boolean =>
+	error instanceof Error && /\(404\)/.test(error.message);
```
[the remark this answers is not in the document]
Comment (KayleeWilliams) [reply packages/cli/src/commands/doctor.ts:125]: Good catch — fixed in ceb2d66. 401/403 from `GET /status/diagnostics` now degrades to `warn` on both `persistence.migrations` and `adapters.health` with an explicit message telling the operator to rerun doctor with an operator/service token (exit 0), instead of surfacing a raw `(403)` hard fail. Other non-404 errors still fail. Covered by a new e2e case where the diagnostics endpoint returns 403 behind a passing auth probe.

**Comment by KayleeWilliams** [comment]
DX follow-up pushed in 378240b: `--help` now renders a **Command flags** section via an optional `CommandDefinition.flagHelp` field (audit tail declares `--request`/`--interval`; doctor needs none since it only uses global flags). Previously, flag-driven commands only showed global flags in help, making required flags like `--request` undiscoverable without the docs. Help snapshots updated.

**Comment by devcool20** [comment]
hey @KayleeWilliams, are you working on this currently?? let me know if any fix is needed!

**Comment by devcool20** [comment]
### Summary of Diagnostics & Test Coverage Fixes

I've resolved the outstanding review issues regarding the `dsar doctor` diagnostics command and backend endpoint:

1. **Fixed TypeScript compiler type mismatch**: Resolved strict TS 6.x build errors inside [webhook-secret-encryption.ts](file:///c:/Users/sharm/OneDrive/Documents/projects/dsar/packages/internals/persistence/src/services/persistence/webhook-secret-encryption.ts) by replacing the type-casting overrides with clean `Uint8Array` casts on Node.js crypto inputs.
2. **Flexible Reachability Schema**: Updated the HTTP diagnostics payload schema in [status.ts (Schema)](file:///c:/Users/sharm/OneDrive/Documents/projects/dsar/packages/backend/src/http-api/groups/status.ts) to define `persistence` as a boolean `reachable` with an optional `error` message string rather than a hardcoded literal `true`.
3. **Graceful Database Unreachability**: Modified [status.ts (Routes)](file:///c:/Users/sharm/OneDrive/Documents/projects/dsar/packages/backend/src/routes/status.ts) to catch errors during database migration queries using `.pipe(Effect.catch(...))`, returning a `reachable: false` diagnostics status instead of crashing the endpoint.
4. **Improved Integration Test Coverage**:
   - Added tests in [status.test.ts](file:///c:/Users/sharm/OneDrive/Documents/projects/dsar/packages/backend/test/routes/status.test.ts) asserting that the diagnostics route handles persistence migration failures and degraded/down adapter states gracefully (returning status 200 with the correct status/errors).
   - Updated the generated OpenAPI snapshot tests to reflect the new dynamic diagnostics contract.
5. **E2E CLI Test Coverage**: Added E2E tests in [doctor.e2e.test.ts](file:///c:/Users/sharm/OneDrive/Documents/projects/dsar/packages/cli/test/e2e/doctor.e2e.test.ts) validating CLI exits with `0` (status `"warn"`) when adapters are degraded, and exits with `1` (status `"fail"`) when adapters are completely down.
6. **Code Documentation**: Added comments in [doctor.ts](file:///c:/Users/sharm/OneDrive/Documents/projects/dsar/packages/cli/src/commands/doctor.ts) documenting the parenthesis-wrapped status error code convention mapped by the `@dsar/guards`'s `toApiError` client utility.
